pkg/config/structure/unmarshal.go

```diff
@@ -0,0 +1,368 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package structure defines a helper to retrieve structured data from the config
+package structure
+
+import (
+	"fmt"
+	"reflect"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/DataDog/datadog-agent/pkg/config/model"
+)
+
+// UnmarshalKey retrieves data from the config at the given key and deserializes it
+// to be stored on the target struct. It is implemented entirely using reflection, and
+// does not depend upon details of the data model of the config.
+// Target struct can use of struct tag of "yaml", "json", or "mapstructure" to rename fields
+func UnmarshalKey(cfg model.Reader, key string, target interface{}) error {
+	source := newNode(reflect.ValueOf(cfg.Get(key)))
+	outValue := reflect.ValueOf(target)
+	if outValue.Kind() == reflect.Pointer {
+		outValue = reflect.Indirect(outValue)
+	}
+	if outValue.Kind() == reflect.Map {
+		return copyMap(outValue, source)
+	}
+	if outValue.Kind() == reflect.Struct {
+		return copyStruct(outValue, source)
+	}
+	if outValue.Kind() == reflect.Slice {
+		return copyList(outValue, source.AsList())
+	}
+	return fmt.Errorf("can only UnmarshalKey to struct or slice, got %v", outValue.Kind())
+}
+
+type nodeType int
+
+const (
+	invalidNodeType nodeType = iota
+	scalarNodeType
+	listNodeType
+	mapNodeType
+	structNodeType
+)
+
+type leafimpl struct {
+	val reflect.Value
+}
+
+type leaf interface {
+	GetBool() (bool, error)
+	GetInt() (int, error)
+	GetString() (string, error)
+}
+
+type nodeListImpl struct {
+	val reflect.Value
+}
+
+type nodeList interface {
+	Size() int
+	Index(int) node
+}
+
+type nodeimpl struct {
+	typ nodeType
+	val reflect.Value
+}
+
+type node interface {
+	GetChild(string) node
+	ChildrenKeys() []string
+	AsList() nodeList
+	AsScalar() leaf
+}
```

Comment:
Why have interface for each type here ? It seems we will only ever have 1 implementation for each one, no ?

Reply:
This question made me realize a better approach for representing and constructing the nodes. Now each node is always the implementation that matches its type (inner node, leaf node, array node, map node) and `newNode` returns that implementation immediately, instead of carrying around the reflect kind (the `typ` member variable) to do so later. This removes the need for `AsScalar` and `AsList`, now we just use interface type assertions instead.

This plan is mirroring how the new config model will work. So the main idea is using interfaces instead of implementations in order to control how the model's internal data is being accessed. Getting type dispatch to handle node behavior is a nice benefit.